File: Pod/Classes/ios/NYTPhotoCaptionView.m
//
//  NYTPhotoCaptionView.m
//  Pods
//
//  Created by Brian Capps on 2/18/15.
//
//

#import "NYTPhotoCaptionView.h"

const CGFloat NYTPhotoCaptionViewHorizontalMargin = 16.0;
static const CGFloat NYTPhotoCaptionViewVerticalMargin = 12.0;

@interface NYTPhotoCaptionView ()

@property (nonatomic) NSAttributedString *attributedTitle;
@property (nonatomic) NSAttributedString *attributedSummary;
@property (nonatomic) NSAttributedString *attributedCredit;

@property (nonatomic) UILabel *textLabel;
@property (nonatomic) CAGradientLayer *gradientLayer;

@end

@implementation NYTPhotoCaptionView

#pragma mark - UIView

- (instancetype)initWithFrame:(CGRect)frame {
    return [self initWithAttributedTitle:nil attributedSummary:nil attributedCredit:nil];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    self.gradientLayer.frame = self.layer.bounds;
}

#pragma mark - NYTPhotoCaptionView

- (instancetype)initWithAttributedTitle:(NSAttributedString *)attributedTitle attributedSummary:(NSAttributedString *)attributedSummary attributedCredit:(NSAttributedString *)attributedCredit {
    self = [super initWithFrame:CGRectZero];
    
    if (self) {
        self.translatesAutoresizingMaskIntoConstraints = NO;
        
        _attributedTitle = attributedTitle;
        _attributedSummary = attributedSummary;
        _attributedCredit = attributedCredit;
        
        [self setupTextLabel];
        [self updateTextLabelAttributedText];
        [self setupGradient];
    }
    
    return self;
}

- (void)setupTextLabel {
    self.textLabel = [[UILabel alloc] init];
    self.textLabel.numberOfLines = 0;
    self.textLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.textLabel];
    
    NSLayoutConstraint *topConstraint = [NSLayoutConstraint constraintWithItem:self.textLabel attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1.0 constant:NYTPhotoCaptionViewVerticalMargin];
    NSLayoutConstraint *bottomConstraint = [NSLayoutConstraint constraintWithItem:self.textLabel attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeBottom multiplier:1.0 constant:-NYTPhotoCaptionViewVerticalMargin];
    NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:self.textLabel attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeWidth multiplier:1.0 constant:-NYTPhotoCaptionViewHorizontalMargin * 2.0];
    NSLayoutConstraint *horizontalPositionConstraint = [NSLayoutConstraint constraintWithItem:self.textLabel attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0.0];
    
    [self addConstraints:@[topConstraint, bottomConstraint, widthConstraint, horizontalPositionConstraint]];
}

- (void)setupGradient {
    self.gradientLayer = [CAGradientLayer layer];
    self.gradientLayer.frame = self.layer.bounds;
    self.gradientLayer.colors = [NSArray arrayWithObjects:(id)[UIColor clearColor].CGColor, (id)[[UIColor blackColor] colorWithAlphaComponent:0.85].CGColor, nil];
    [self.layer insertSublayer:self.gradientLayer atIndex:0];
}

- (void)updateTextLabelAttributedText {
    NSMutableAttributedString *attributedLabelText = [[NSMutableAttributedString alloc] init];
    
    if (self.attributedTitle) {
        [attributedLabelText appendAttributedString:self.attributedTitle];
    }
    
    if (self.attributedSummary) {
        if (self.attributedTitle) {
            [attributedLabelText appendAttributedString:[[NSAttributedString alloc] initWithString:@"\n" attributes:nil]];
        }
        
        [attributedLabelText appendAttributedString:self.attributedSummary];
    }
    
    if (self.attributedCredit) {
        if (self.attributedTitle || self.attributedSummary) {
            [attributedLabelText appendAttributedString:[[NSAttributedString alloc] initWithString:@"\n" attributes:nil]];
        }
        
        [attributedLabelText appendAttributedString:self.attributedCredit];
    }
    
    self.textLabel.attributedText = attributedLabelText;
}

@end

File: Sources/NYTPhotoViewer/NYTPhotoCaptionView.swift
//
//  NYTPhotoCaptionView.swift
//  NYTPhotoViewer
//
//  Created by Brian Capps on 2/18/15.
//  Copyright (c) 2015 NYTimes. All rights reserved.
//

import UIKit

private let horizontalMargin: CGFloat = 8.0
private let verticalMargin: CGFloat = 7.0

/// A view used to display the caption for a photo.
///
/// This is used by default when no custom caption view is provided.
@MainActor
public class NYTPhotoCaptionView: UIView, NYTPhotoCaptionViewLayoutWidthHinting {
    
    // MARK: - Properties
    
    public let attributedTitle: NSAttributedString?
    public let attributedSummary: NSAttributedString?
    public let attributedCredit: NSAttributedString?
    
    private let textView: UITextView
    private let gradientLayer: CAGradientLayer
    
    public var preferredMaxLayoutWidth: CGFloat = 0 {
        didSet {
            let newValue = ceil(preferredMaxLayoutWidth)
            if abs(oldValue - newValue) > 0.1 {
                invalidateIntrinsicContentSize()
            }
        }
    }
    
    // MARK: - Initialization
    
    /// Designated initializer that takes all the caption attributed strings as arguments.
    ///
    /// - Parameters:
    ///   - attributedTitle: The attributed string used as the title. The top string in the caption view.
    ///   - attributedSummary: The attributed string used as the summary. The second from the top string in the caption view.
    ///   - attributedCredit: The attributed string used as the credit. The third from the top string in the caption view.
    public init(attributedTitle: NSAttributedString?, attributedSummary: NSAttributedString?, attributedCredit: NSAttributedString?) {
        self.attributedTitle = attributedTitle
        self.attributedSummary = attributedSummary
        self.attributedCredit = attributedCredit
        self.textView = UITextView(frame: .zero, textContainer: nil)
        self.gradientLayer = CAGradientLayer()
        
        super.init(frame: .zero)
        
        commonInit()
    }
    
    required init?(coder: NSCoder) {
        self.attributedTitle = nil
        self.attributedSummary = nil
        self.attributedCredit = nil
        self.textView = UITextView(frame: .zero, textContainer: nil)
        self.gradientLayer = CAGradientLayer()
        
        super.init(coder: coder)
        
        commonInit()
    }
    
    private func commonInit() {
        translatesAutoresizingMaskIntoConstraints = false
        
        setupTextView()
        updateTextViewAttributedText()
        setupGradient()
    }
    
    // MARK: - Setup
    
    private func setupTextView() {
        textView.translatesAutoresizingMaskIntoConstraints = false
        textView.isEditable = false
        textView.dataDetectorTypes = []
        textView.backgroundColor = .clear
        textView.textContainerInset = UIEdgeInsets(top: verticalMargin, left: horizontalMargin, bottom: verticalMargin, right: horizontalMargin)
        
        addSubview(textView)
        
        NSLayoutConstraint.activate([
            textView.topAnchor.constraint(equalTo: topAnchor),
            textView.bottomAnchor.constraint(equalTo: bottomAnchor),
            textView.widthAnchor.constraint(equalTo: widthAnchor),
            textView.centerXAnchor.constraint(equalTo: centerXAnchor)
        ])
    }
    
    private func setupGradient() {
        gradientLayer.frame = layer.bounds
        gradientLayer.colors = [UIColor.clear.cgColor, UIColor.black.withAlphaComponent(0.85).cgColor]
        layer.insertSublayer(gradientLayer, at: 0)
    }
    
    private func updateTextViewAttributedText() {
        let attributedLabelText = NSMutableAttributedString()
        
        if let title = attributedTitle {
            attributedLabelText.append(title)
        }
        
        if let summary = attributedSummary {
            if attributedTitle != nil {
                attributedLabelText.append(NSAttributedString(string: "\n"))
            }
            attributedLabelText.append(summary)
        }
        
        if let credit = attributedCredit {
            if attributedTitle != nil || attributedSummary != nil {
                attributedLabelText.append(NSAttributedString(string: "\n"))
            }
            attributedLabelText.append(credit)
        }
        
        textView.attributedText = attributedLabelText
    }
    
    // MARK: - UIView Overrides
    
    public override class var requiresConstraintBasedLayout: Bool {
        return true
    }
    
    public override func didMoveToSuperview() {
        super.didMoveToSuperview()
        
        guard let superview = superview else { return }
        
        NSLayoutConstraint.activate([
            heightAnchor.constraint(lessThanOrEqualTo: superview.heightAnchor, multiplier: 0.3)
        ])
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        
        func updateGradientFrame() {
            if #available(iOS 11.0, *) {
                let safeAreaInsets = window?.safeAreaInsets ?? .zero
                let selfBounds = layer.bounds
                gradientLayer.frame = CGRect(
                    x: selfBounds.origin.x - safeAreaInsets.left,
                    y: selfBounds.origin.y + safeAreaInsets.bottom,
                    width: selfBounds.size.width + safeAreaInsets.left + safeAreaInsets.right,
                    height: selfBounds.size.height + safeAreaInsets.bottom
                )
            } else {
                gradientLayer.frame = layer.bounds
            }
        }
        
        updateGradientFrame()
        
        // On iOS 8.x, when this view is height-constrained, neither `self.bounds` nor `self.layer.bounds` reflects the new layout height immediately after `layoutSubviews`. Both of those properties appear correct in the next runloop.
        DispatchQueue.main.async {
            updateGradientFrame()
        }
    }
    
    public override var intrinsicContentSize: CGSize {
        let contentSize = textView.sizeThatFits(CGSize(width: preferredMaxLayoutWidth, height: .greatestFiniteMagnitude))
        let width = preferredMaxLayoutWidth
        let height = ceil(contentSize.height)
        
        return CGSize(width: width, height: height)
    }
}

